pkg/controller/projects/deploykeys/controller.go

```diff
@@ -0,0 +1,274 @@
+	errors "github.com/pkg/errors"
+	gitlab "github.com/xanzy/go-gitlab"
+
+	kubeCoreV1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kubeTypes "k8s.io/apimachinery/pkg/types"
+	pointer "k8s.io/utils/pointer"
+	kubeController "sigs.k8s.io/controller-runtime"
+	kubeClient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	v1alpha1 "github.com/crossplane-contrib/provider-gitlab/apis/projects/v1alpha1"
+	clients "github.com/crossplane-contrib/provider-gitlab/pkg/clients"
+	projects "github.com/crossplane-contrib/provider-gitlab/pkg/clients/projects"
+)
+
+const (
+	errNotDeployKey     = "managed resource is not a Gitlab deploy key custom resource"
+	errNotFound         = "404 project deploy key not found"
+	errGetFail          = "cannot get Gitlab deploy key"
+	errCreateFail       = "cannot create Gitlab deploy key"
+	errUpdateFail       = "cannot update Gitlab deploy key"
+	errDeleteFail       = "cannot delete Gitlab deploy key"
+	errKeyMissing       = "missing key ref value"
+	errIDNotAnInt       = "external-name is not an int"
+	errProjectIDMissing = "missing project ID"
+)
+
+type external struct {
+	kube   kubeClient.Client
+	client projects.DeployKeyClient
+}
+
+type connector struct {
+	kube              kubeClient.Client
+	newGitlabClientFn func(clientConfig clients.Config) projects.DeployKeyClient
+}
+
+// SetupDeployKey adds a controller that reconciles ProjectDeployKey.
+func SetupDeployKey(manager kubeController.Manager, log logging.Logger) error {
+	name := managed.ControllerName(v1alpha1.DeployKeyKind)
+
+	connector := &connector{kube: manager.GetClient(), newGitlabClientFn: newDeployKeyClient}
+
+	reconciler := managed.NewReconciler(manager,
+		resource.ManagedKind(v1alpha1.DeployKeyGroupVersionKind),
+		managed.WithExternalConnecter(connector),
+		managed.WithInitializers(managed.NewDefaultProviderConfig(manager.GetClient())),
+		managed.WithLogger(log.WithValues("controller", name)),
+		managed.WithRecorder(event.NewAPIRecorder(manager.GetEventRecorderFor(name))))
+
+	return kubeController.NewControllerManagedBy(manager).
+		Named(name).
+		For(&v1alpha1.DeployKey{}).
+		Complete(reconciler)
+}
+
+func (c *connector) Connect(ctx context.Context, mgd resource.Managed) (managed.ExternalClient, error) {
+	cr, ok := mgd.(*v1alpha1.DeployKey)
+
+	if !ok {
+		return nil, errors.New(errNotDeployKey)
+	}
+
+	config, err := clients.GetConfig(ctx, c.kube, cr)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &external{kube: c.kube, client: c.newGitlabClientFn(*config)}, nil
+}
+
+func (e *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mg.(*v1alpha1.DeployKey)
+
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotDeployKey)
+	}
+
+	if cr.Spec.ForProvider.ProjectID == nil {
+		return managed.ExternalObservation{}, errors.New(errProjectIDMissing)
+	}
+
+	ID, err := strconv.Atoi(meta.GetExternalName(cr))
```

Comment:
Local variables should be lower case.

```suggestion
	id, err := strconv.Atoi(meta.GetExternalName(cr))
```